### DISCUSSION TITLE
A good starter .golangci.yml  ?

### DISCUSSION THREAD

**Discussion by bytebeast** [golangci/golangci-lint#4393] [Q&A] [answered]
What is a good default linter set to install with a good default .golangci.yml to use with it?

**Comment by bombsimon** [comment, marked as answer]
I don't think there's any one size fits all here. If anything I would say an empty config is the best place to start. But I see a few approaches you can take.

## Just the bare minimum

This is what I mean with default. You can create an empty `.golanci.yaml` so it's easy to change things when you notice what it does for you, but you don't have to. Maybe you just want to start with some formatting options or you don't want to limit the number of same issues? Then this would be enough:

```yaml
---
run:
  tests: true

output:
  format: colored-line-number
  print-issued-lines: false

issues:
  max-issues-per-linter: 0
  max-same-issues: 0
```

## Try it all

You can take a different approach and just enable everything and see where it fails. It might be a good idea for new projects but will probably very hard to implement for existing bigger codebases. Just enable everything and disable all exclusions. If you get annoyed by the warnings of deprecated linters you can disable them. Something like this:

```yaml
---
run:
  deadline: 1m

linters:
  enable-all: true

issues:
  exclude-use-default: false
```

## Find someone you trust

Just find a project that you think has a good coding standard or comes from a user or organization that you trust. F.ex. [GitHub's `gh-ost` configuration](https://github.com/github/gh-ost/blob/master/.golangci.yml) or maybe [how traefik does it](https://github.com/traefik/traefik/blob/master/.golangci.yml) is what you want. Just copy that one and go from there.

## Just take the time to read up on linters

All linters are fairly [well documented](https://golangci-lint.run/usage/linters/) and contains a description of what it does, a category to tell if it's stylistic, security, to find bugs etc. Just read the list and look for things you care about. You usually don't have to configure any of them from scratch and can leave that until you find potential issues or things you don't like with a linter. If you can't configure it to your likings you can disable it in the future.

The rest of the configuration is also [documented here](https://golangci-lint.run/usage/configuration/) and even though there's a lot to read up on just starting at one block and going from there is what you like.

If you find your way to `golangci-lint` chances are that you're somewhat familiar with different linters and formatters out there and have specific wishes in mind.

---

If you want to contribute to evolving different approaches to presets or different onboardings feel free to comment in #1141 or #3954 (there's probably more). Just keep in mind that `golangci-lint` is not _one_ linter like `eslint`, it's a meta linter bundling a ton of linters so it's even harder to recommend sensible defaults since the linters comes from different authors at different times with different needs and goals in mind.
